src/sinks/sematext/metrics.rs

```diff
@@ -0,0 +1,379 @@
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "sematext_metrics"
+    }
+}
+
+lazy_static! {
+    static ref REQUEST_DEFAULTS: TowerRequestConfig = TowerRequestConfig {
+        retry_attempts: Some(5),
+        ..Default::default()
+    };
+}
+
+fn write_uri(endpoint: &str) -> crate::Result<Uri> {
+    encode_uri(
+        &endpoint,
+        "write",
+        &[
+            ("db", Some("metrics".into())),
+            ("v", Some(format!("vector-{}", vector_version()))),
+            ("precision", Some("ns".into())),
+        ],
+    )
+}
+
+impl SematextMetricsSvc {
+    pub fn new(
+        config: SematextMetricsConfig,
+        endpoint: http::Uri,
+        cx: SinkContext,
+        client: HttpClient,
+    ) -> crate::Result<VectorSink> {
+        let batch = BatchSettings::default()
+            .events(20)
+            .timeout(1)
+            .parse_config(config.batch)?;
+        let request = config.request.unwrap_with(&REQUEST_DEFAULTS);
+        let http_service = HttpBatchService::new(client, create_build_request(endpoint));
+        let sematext_service = SematextMetricsSvc {
+            config,
+            inner: http_service,
+        };
+
+        let sink = request
+            .batch_sink(
+                HttpRetryLogic,
+                sematext_service,
+                MetricBuffer::new(batch.size),
+                batch.timeout,
+                cx.acker(),
+            )
+            .sink_map_err(|e| error!("Fatal sematext metrics sink error: {}", e));
+
+        Ok(VectorSink::Futures01Sink(Box::new(sink)))
+    }
+}
+
+impl Service<Vec<Metric>> for SematextMetricsSvc {
+    type Response = http::Response<bytes::Bytes>;
+    type Error = crate::Error;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut std::task::Context) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, items: Vec<Metric>) -> Self::Future {
+        let input = encode_events(&self.config.token, items);
+        let body: Vec<u8> = input.into_bytes();
+
+        self.inner.call(body)
+    }
+}
+
+fn create_build_request(
+    uri: http::Uri,
+) -> impl Fn(Vec<u8>) -> BoxFuture<'static, crate::Result<hyper::Request<Vec<u8>>>> + Sync + Send + 'static
+{
+    move |body| {
+        Box::pin(ready(
+            hyper::Request::post(uri.clone())
+                .header("Content-Type", "text/plain")
+                .body(body)
+                .map_err(Into::into),
+        ))
+    }
+}
+
+/// Sematext takes the first part of the name as the namespace for the event.
+/// The rest of the name is the value label.
+/// If the name is not a '.' separated string, it will take the whole name as the
+/// namespace, and set label the value as 'value'.
+/// This probably should not happen if you want meaningful metrics in Sematext.
```

Comment:
This is a very good question. 

Sematext metrics works with a set number of specific labels. In order to use this sink with Sematext effectively, you will need to know which metrics you are going to be sending into Sematext. If you did use a metric name that didn't follow the `namespace.value` format, then that metric wouldn't really show up anywhere in Sematext anyway. So I think we probably do not need to document this. @binarylogic what do you think?

Reply:
Does https://github.com/timberio/vector/pull/3686 help with this (cc @jszwedko)?

I may be misunderstanding this. In general, our internal metrics data model should have enough structure and meaning to transform the data into the integration's preferred shape. We should not require brittle naming rules for it to work. https://github.com/timberio/vector/pull/3686 is a good example of adjusting our data model to achieve this. And #3910 proposes something similar for the logging domain.

Does that help?

Reply:
Can you throw some examples of this in here? Like an incoming metric is x and the outputted Semtatext-friendly metric is y? Then let's look at user-facing documentation explaining the incoming/outcoming scenario.

Reply:
It's more to be inline with Sematexts naming rules that they have designed their dashboards around. So, for example in Sematext they have a JVM dashboard that tracks various JVM parameters. To send the garbage collection time, you would send a metric with the label 'jvm.gc.collection.time', number of threads is `jvm.threads` and so on.

Reply:
[Here](https://sematext.com/docs/synthetics/metrics/#custom-metrics) is an example from their documentation with some of the metrics they collect.

Reply:
I think this merits some user-facing docs and then we can discuss other, larger-scale encoding/codecs/etc in the other tickets @binarylogic has open.

Reply:
Cool. I think it will also be well worth running any docs we do past the team at Sematext for a sanity check. 

Do you want me to have a go at these docs?

Reply:
If you could please! Feel free to ping me and I can help too. And yes - we should ensure the Sematext team is okay. Thanks!

Reply:
@jamtur01 It turns out that I was mistaken. I've confirmed with Sematext that actually there are no restrictions on the naming for labels. There is a bug in their code that meant if you send a custom metric without any tags it doesn't show up in their reports - which explains why I wasn't getting it working. That is going to be fixed this week.

So, I think the docs probably just need to make clear that the sink only accepts Counter and Gauge metrics and it should be good to go.

Reply:
Have added some documentation [here](https://github.com/timberio/vector-website/pull/126).

Reply:
@binarylogic Given this is just waiting for the docs, which I can sort out as part of the migration to cue, do you think this pull request is good to merge?